### PR TITLE
Windows Support

### PR DESCRIPTION
**Pull request by KingRial**
Since I loved the library, I just added a quick support for windows based on [WMI](https://docs.microsoft.com/en-us/windows/win32/wmisdk/querying-wmi) queries.

I am pretty new to GoLang so, this is far from being perfect or fully completed, but at least it's a start.

Hope it could be of any help.

### CODE REVIEW

**Comment by jaypipes** [comment]
Hi! Thank you so much @KingRial for this PR! Sorry for the delay in reviewing, was on PTO... will review in next day or so and test out on a Windows VM. Thanks! :)

**Comment by KingRial** [comment]
Hello @jaypipes!
It is my pleasure to collaborate with this project.

A quick overview of what I am offering:

- A support for **windows HW inspection**
I still know there are big area of improvements but, at least, there are the main informations and anyone can improve them
- **Memory struct**
added how many phsysical "memory banks" are present and their main informations
(let me know if you prefer a different way to handle them)

As an inspection library I presumed it could be interesting in having also some non-HW data.
**N.B.** If it distorts the library philosophy I will remove them!

- **NIC struct**
started to add some SW configuration: IP / Gateways / DHCP
- **Packages**
Added the ability to list the installed packages/softwares on the HW (right now it supports windows only but adding the linux/darwin support is just a matter of time)

Let me know if  can help you somehow; but please remember I am pretty new to GoLang ^^

**Comment by jaypipes** [comment]
Hello again, @KingRial! Just checking in with you. I'm writing this from a Windows machine where I am testing out your patch and doing a full code review. I'm also reviewing the github.com/StackExchange/wmi library at the same time to evaluate including it as a dependency for ghw.

Would it be possible to separate your pull request into two individual PRs, the first being the changes to add Windows support without any additional fields to the core `ghw` API structs (such as the new `MemoryInfo.MemoryBanks` field), the second being any changes to the core `ghw` structs?

File: baseboard_windows.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	Manufacturer string
+	SerialNumber string
+	Tag          string
+	Version      string
+}
+
+func (ctx *context) baseboardFillInfo(info *BaseboardInfo) error {
+	// Getting disk drives from WMI
+	var win32BaseboardDescriptions []Win32_BaseBoard
+	q1 := wmi.CreateQuery(&win32BaseboardDescriptions, "")
```
Comment (jaypipes) [inline baseboard_windows.go]: You don't need to use the `wmi.CreateQuery()` function here. Instead, simply create a const string with the WQL query, like so:

```go
const (
        wqlBaseboard = "SELECT Manufacturer, SerialNumber, Tag, Version FROM Win32_BaseBoard"
)

type win32Baseboard struct {
	Manufacturer string
	SerialNumber string
	Tag          string
	Version      string
}

func (ctx *context) baseboardFillInfo(info *BaseboardInfo) error {
        var win32BaseboardDescriptions []win32Baseboard
        if err := wmi.Query(wqlBaseboard, &win32BaseboardDescriptions); err != nil {
		return err
	}
	if len(win32BaseboardDescriptions) > 0 {
		info.AssetTag = win32BaseboardDescriptions[0].Manufacturer
		info.SerialNumber = win32BaseboardDescriptions[0].Manufacturer
		info.Vendor = win32BaseboardDescriptions[0].Manufacturer
		info.Version = win32BaseboardDescriptions[0].Manufacturer
	}
	return nil
}
```

File: bios_windows.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// Use and distribution licensed under the Apache license version 2.
+//
+// See the COPYING file in the root project directory for full text.
+//
+
+package ghw
+
+import "github.com/StackExchange/wmi"
+
+type CIM_BIOSElement struct {
```
Comment (jaypipes) [inline bios_windows.go]: Same comment about naming of the struct. Please use `biosElement` and just pass a const string of `SELECT InstallDate, Manufacturer, Version FROM CMI_BIOSElement` to `cmi.Query()` below.

File: bios_windows.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	InstallDate  string
+	Manufacturer string
+	Version      string
+}
+
+func (ctx *context) biosFillInfo(info *BIOSInfo) error {
+	// Getting disk drives from WMI
```
Comment (jaypipes) [inline bios_windows.go]: just remove the above comment :)

File: block.go [hunk 4]
```diff
@@ -108,13 +108,14 @@ type Disk struct {
...
-	MountPoint string `json:"mount_point"`
-	SizeBytes  uint64 `json:"size_bytes"`
-	Type       string `json:"type"`
-	IsReadOnly bool   `json:"read_only"`
+	Disk        *Disk  `json:"-"`
+	Name        string `json:"name"`
+	Label       string `json:"label"`
+	MountPoint  string `json:"mount_point"`
+	SizeBytes   uint64 `json:"size_bytes"`
+	UsableBytes uint64 `json:"usable_bytes"`
```
Comment (jaypipes) [inline block.go]: Please pull this change into a separate pull request :)

File: block_windows.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+func toReadOnly(access uint16) bool {
+	var readOnly bool
+	switch access {
+	case 0x1: // See Access property from: https://docs.microsoft.com/en-us/windows/win32/cimwin32prov/win32-diskpartition
+		readOnly = true
+	default:
+		readOnly = false
+	}
+	return readOnly
+}
```
Comment (jaypipes) [inline block_windows.go]: The above function can be simplified to just this:

```go
func toReadOnly(access uint16) bool {
	// See Access property from: https://docs.microsoft.com/en-us/windows/win32/cimwin32prov/win32-diskpartition
        return access == 0x1
}
```

File: block_windows.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+								IsReadOnly:  toReadOnly(diskpartition.Access), // TODO: add information
+							}
+							disk.Partitions = append(disk.Partitions, p)
+							break
+						}
+					}
+				}
+			}
+		}
+		// Appending Disk
```
Comment (jaypipes) [inline block_windows.go]: you can remove the above comment :)

File: block_windows.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+						//fmt.Printf("diskpartition %#v\n", diskpartition)
+						var desiredAntecedent = "\\\\" + diskpartition.SystemName + "\\root\\cimv2:" + diskpartition.CreationClassName + ".DeviceID=\"" + diskpartition.DeviceID + "\""
+						var desiredDependent = "\\\\" + logicaldisk.SystemName + "\\root\\cimv2:" + logicaldisk.CreationClassName + ".DeviceID=\"" + logicaldisk.DeviceID + "\""
+						//fmt.Printf("-- Antecedent\n%#v\n%#v\n", desiredAntecedent, logicaldisktodiskpartition.Antecedent)
+						//fmt.Printf("-- Dependent\n%#v\n%#v\n", desiredDependent, logicaldisktodiskpartition.Dependent)
+						if logicaldisktodiskpartition.Antecedent == desiredAntecedent && logicaldisktodiskpartition.Dependent == desiredDependent {
+							//fmt.Printf("--------------- Disk drive %#v\n", diskdrive)
+							//fmt.Printf("--------------- Disk Partition %#v\n", diskpartition)
+							//fmt.Printf("--------------- Logical Disk %#v\n", logicaldisk)
+							// Appending Partition
```
Comment (jaypipes) [inline block_windows.go]: you can remove the above comment

File: block_windows.go [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+						//fmt.Printf("logicaldisk %#v\n", logicaldisk)
+						//fmt.Printf("diskpartition %#v\n", diskpartition)
+						var desiredAntecedent = "\\\\" + diskpartition.SystemName + "\\root\\cimv2:" + diskpartition.CreationClassName + ".DeviceID=\"" + diskpartition.DeviceID + "\""
+						var desiredDependent = "\\\\" + logicaldisk.SystemName + "\\root\\cimv2:" + logicaldisk.CreationClassName + ".DeviceID=\"" + logicaldisk.DeviceID + "\""
+						//fmt.Printf("-- Antecedent\n%#v\n%#v\n", desiredAntecedent, logicaldisktodiskpartition.Antecedent)
+						//fmt.Printf("-- Dependent\n%#v\n%#v\n", desiredDependent, logicaldisktodiskpartition.Dependent)
+						if logicaldisktodiskpartition.Antecedent == desiredAntecedent && logicaldisktodiskpartition.Dependent == desiredDependent {
+							//fmt.Printf("--------------- Disk drive %#v\n", diskdrive)
+							//fmt.Printf("--------------- Disk Partition %#v\n", diskpartition)
+							//fmt.Printf("--------------- Logical Disk %#v\n", logicaldisk)
```
Comment (jaypipes) [inline block_windows.go]: Please remove above commented-out prints.

File: block_windows.go [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+				// Finding logical partition linked to current disk partition
+				for _, logicaldisk := range win32LogicalDiskDescriptions {
+					for _, logicaldisktodiskpartition := range win32LogicalDiskToPartitionDescriptions {
+						//fmt.Printf("\nRelation %#v\n", logicaldisktodiskpartition)
+						//fmt.Printf("logicaldisk %#v\n", logicaldisk)
+						//fmt.Printf("diskpartition %#v\n", diskpartition)
+						var desiredAntecedent = "\\\\" + diskpartition.SystemName + "\\root\\cimv2:" + diskpartition.CreationClassName + ".DeviceID=\"" + diskpartition.DeviceID + "\""
+						var desiredDependent = "\\\\" + logicaldisk.SystemName + "\\root\\cimv2:" + logicaldisk.CreationClassName + ".DeviceID=\"" + logicaldisk.DeviceID + "\""
+						//fmt.Printf("-- Antecedent\n%#v\n%#v\n", desiredAntecedent, logicaldisktodiskpartition.Antecedent)
+						//fmt.Printf("-- Dependent\n%#v\n%#v\n", desiredDependent, logicaldisktodiskpartition.Dependent)
```
Comment (jaypipes) [inline block_windows.go]: Please remove above commented-out prints.

File: block_windows.go [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+			// Finding disk partition linked to current disk drive
+			if diskdrive.Index == diskpartition.DiskIndex {
+				disk.PhysicalBlockSizeBytes = diskpartition.BlockSize
+				//fmt.Printf("Disk Partition %#v\n", diskdrive)
+				// Finding logical partition linked to current disk partition
+				for _, logicaldisk := range win32LogicalDiskDescriptions {
+					for _, logicaldisktodiskpartition := range win32LogicalDiskToPartitionDescriptions {
+						//fmt.Printf("\nRelation %#v\n", logicaldisktodiskpartition)
+						//fmt.Printf("logicaldisk %#v\n", logicaldisk)
+						//fmt.Printf("diskpartition %#v\n", diskpartition)
```
Comment (jaypipes) [inline block_windows.go]: Please remove above commented-out prints.

File: block_windows.go [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+			Model:        diskdrive.Caption,
+			SerialNumber: diskdrive.SerialNumber,
+			WWN:          UNKNOWN, // TODO: add information
+			Partitions:   make([]*Partition, 0),
+		}
+		for _, diskpartition := range win32DiskPartitionDescriptions {
+			// Finding disk partition linked to current disk drive
+			if diskdrive.Index == diskpartition.DiskIndex {
+				disk.PhysicalBlockSizeBytes = diskpartition.BlockSize
+				//fmt.Printf("Disk Partition %#v\n", diskdrive)
```
Comment (jaypipes) [inline block_windows.go]: Please remove above commented-out prints.

File: block_windows.go [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+	for _, diskdrive := range win32DiskDriveDescriptions {
+		disk := &Disk{
+			Name:                   diskdrive.Name,
+			SizeBytes:              diskdrive.Size,
+			PhysicalBlockSizeBytes: 0,
+			DriveType:              toDriveType(diskdrive.MediaType),
+			StorageController:      toStorageController(diskdrive.InterfaceType),
+			BusType:                toBusType(diskdrive.InterfaceType),
+			BusPath:                UNKNOWN, // TODO: add information
+			//NUMANodeID:             node,
```
Comment (jaypipes) [inline block_windows.go]: Please remove above. NUMANodeID isn't a field of Disk :)

File: block_windows.go [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+	if err := wmi.Query(q3, &win32LogicalDiskToPartitionDescriptions); err != nil {
+		return err
+	}
+
+	// Getting logical disks from WMI
+	var win32LogicalDiskDescriptions []Win32_LogicalDisk
+	q4 := wmi.CreateQuery(&win32LogicalDiskDescriptions, "")
+	if err := wmi.Query(q4, &win32LogicalDiskDescriptions); err != nil {
+		return err
+	}
```
Comment (jaypipes) [inline block_windows.go]: for each of the above, please create a separate helper function that returns the slice of appropriate structs, where the helper function uses a const string `cmi.Query()` instead of the `cmi.CreateQuery()` function.

Example for the disk drives:

```go
const (
        wqlDiskDrives = `
SELECT Caption, CreationClassName, Description, DeviceID,
    Index, InterfaceType. Manufacturer, MediaType, Model, Name,
    SerialNumber, Size, TotalCylinders, TotalHeads, TotalSectors,
    TotalTracks, TracksPerCylinder
FROM Win32_DiskDrive`
)

func (ctx *context) blockFillInfo(info *BlockInfo) error {
        win32DiskDriveDescriptions, err := getDiskDrives()
        if err != nil {
                return err
        }
...
}

func getDiskDrives() ([]win32DiskDrive, error) {
	// Getting disk drives from WMI
	var win32DiskDriveDescriptions []win32DiskDrive
	if err := wmi.Query(wqlDiskDrives, &win32DiskDriveDescriptions); err != nil {
		return nil, err
	}
        return win3232DiskDriveDescriptions, nil
}
```

File: block_windows.go [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+	Size              uint64
+	TotalCylinders    int64
+	TotalHeads        int32
+	TotalSectors      int64
+	TotalTracks       int64
+	TracksPerCylinder int32
+}
+
+/*type Win32_DiskDriveToDiskPartition struct {
+}*/
```
Comment (jaypipes) [inline block_windows.go]: unused above. please remove.

File: chassis_windows.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+// See the COPYING file in the root project directory for full text.
+//
+
+package ghw
+
+import (
+	"github.com/StackExchange/wmi"
+)
+
+type CIM_Chassis struct {
```
Comment (jaypipes) [inline chassis_windows.go]: Same comment as above about using an unexported struct name and a const string for the WQL query.

File: cpu_windows.go [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	return nil
+}
+
+func (ctx *context) processorsGet(win32descriptions []Win32_Processor) []*Processor {
+	var procs []*Processor
+	// Converting into standard structures
+	for index, description := range win32descriptions {
+		p := &Processor{
+			Id:         index,            // TODO: how to get a decent "Physical ID" to use ?
+			Model:      description.Name, // description.ProcessorID description.Manufacturer description.Name
```
Comment (jaypipes) [inline cpu_windows.go]: please remove comment at end of this line.

File: cpu_windows.go [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+// See the COPYING file in the root project directory for full text.
+//
+
+package ghw
+
+import (
+	"github.com/StackExchange/wmi"
+)
+
+type Win32_Processor struct {
```
Comment (jaypipes) [inline cpu_windows.go]: same comment about the struct name and using a const query string

File: gpu_windows.go [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+// See the COPYING file in the root project directory for full text.
+//
+
+package ghw
+
+import (
+	"github.com/StackExchange/wmi"
+)
+
+type Win32_VideoController struct {
```
Comment (jaypipes) [inline gpu_windows.go]: same comment about the struct name and using a const query string

File: gpu_windows.go [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+	for _, description := range win32VideoControllerDescriptions {
+		card := &GraphicsCard{
+			Address:    description.DeviceID, // https://stackoverflow.com/questions/32073667/how-do-i-discover-the-pcie-bus-topology-and-slot-numbers-on-the-board
+			Index:      0,
+			DeviceInfo: pci.GetDevice(description.PNPDeviceID),
+		}
+		cards = append(cards, card)
+	}
+	//car.DeviceInfo
+	//card.Node
```
Comment (jaypipes) [inline gpu_windows.go]: Please remove the above comments

File: memory.go [hunk 20]
```diff
@@ -11,9 +11,19 @@ import (
...
 )
 
+type MemoryBank struct {
+	Name         string `json:"name"`
+	Label        string `json:"label"`
+	Location     string `json:"location"`
+	SerialNumber string `json:"vendor"`
+	SizeBytes    int64  `json:"size_bytes"`
+	Vendor       string `json:"vendor"`
+}
```
Comment (jaypipes) [inline memory.go]: Please separate these changes into a different pull request.

File: memory_windows.go [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+// See the COPYING file in the root project directory for full text.
+//
+
+package ghw
+
+import (
+	"github.com/StackExchange/wmi"
+)
+
+type Win32_OperatingSystem struct {
```
Comment (jaypipes) [inline memory_windows.go]: same comment about the struct name and using a const query string

File: net.go [hunk 22]
```diff
@@ -10,17 +10,26 @@ import (
 	"fmt"
 )
 
+type NICConfiguration struct {
+	DHCPenabled bool   `json:"dhcp_enabled"`
+	Gateway     string `json:"gateway"`
+	IPv4        string `json:"ipv4"`
+	IPv6        string `json:"ipv6"`
+}
```
Comment (jaypipes) [inline net.go]: Please move the changes adding this new NICConfiguration to a separate pull request where we can discuss this feature in isolation from the Windows-specific patch here.

File: net_windows.go [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+	q3 := wmi.CreateQuery(&win32IP4RouteTableDescriptions, "")
+	if err := wmi.Query(q3, &win32IP4RouteTableDescriptions); err != nil {
+		return err
+	}
+
+	info.NICs = ctx.nics(win32NetDescriptions, win32NetConfigurationDescriptions, win32IP4RouteTableDescriptions)
+	return nil
+}
+
+type Win32_NetworkAdapter struct {
```
Comment (jaypipes) [inline net_windows.go]: same comment about the struct name and using a const query string

File: packages.go [hunk 24]
```diff
@@ -0,0 +1,68 @@
+//
```
Comment (jaypipes) [inline packages.go]: Please move all the additional packages module to a new pull request. It is not related to the Windows PR or the NIC Configuration or memory banks work.

File: product_windows.go [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+// See the COPYING file in the root project directory for full text.
+//
+
+package ghw
+
+import (
+	"github.com/StackExchange/wmi"
+)
+
+type Win32_ComputerSystemProduct struct {
```
Comment (jaypipes) [inline product_windows.go]: same comment about the struct name and using a const query string

File: pci_windows.go [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+			ID:                    UNKNOWN,
+			Name:                  UNKNOWN,
+			ProgrammingInterfaces: []*pcidb.ProgrammingInterface{},
+		},
+		ProgrammingInterface: &pcidb.ProgrammingInterface{
+			ID:   UNKNOWN,
+			Name: UNKNOWN,
+		},
+	}
+	// Getting disk drives from WMI
```
Comment (jaypipes) [inline pci_windows.go]: Not getting disk drives :)

File: pci_windows.go [hunk 27]
```diff
@@ -0,0 +1,87 @@
+// Use and distribution licensed under the Apache license version 2.
+//
+// See the COPYING file in the root project directory for full text.
+//
```
Comment (jaypipes) [inline pci_windows.go]: Let's leave PCI information on Windows to a separate pull request, OK? I need to do some more research on just how relevant the PnPEntity thing in WMI is to a PCI device. Please just pull all the PCI code out into a separate pull request.

**Comment by jaypipes** [comment]
@KingRial can you email me at jaypipes at gmail dot com? Wanted to check with you about something privately... thanks!

File: baseboard_windows.go [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+
+type Win32_BaseBoard struct {
+	Manufacturer string
+	SerialNumber string
+	Tag          string
+	Version      string
+}
+
+func (ctx *context) baseboardFillInfo(info *BaseboardInfo) error {
+	// Getting disk drives from WMI
```
[the remark this answers is not in the document]
Comment (KingRial) [reply baseboard_windows.go]: My bad!

File: baseboard_windows.go [hunk 29]
```diff
@@ -0,0 +1,32 @@
+// Use and distribution licensed under the Apache license version 2.
+//
+// See the COPYING file in the root project directory for full text.
+//
+
+package ghw
+
+import "github.com/StackExchange/wmi"
+
+type Win32_BaseBoard struct {
```
[the remark this answers is not in the document]
Comment (KingRial) [reply baseboard_windows.go]: I agree on changing to unexported struct

File: baseboard_windows.go (hunk 1 above)
Comment (KingRial) [reply baseboard_windows.go]: That is true.

The main reason I was using the same pattern was because I could just change the **win32Baseboard** struct obtaining dynamically the correct updated query.

Pretty usefull if you are lazy like me and you are always exploring WMI tables to find the correct information ^^

Let me know if you still think it's better to use a static query instead of having a dynamic one built from the structure.

File: bios_windows.go (hunk 3 above)
Comment (KingRial) [reply bios_windows.go]: Done ^^

File: pci_windows.go (hunk 27 above)
Comment (KingRial) [reply pci_windows.go]: Right now GPU e PCI are linked to try to follow the same pattern used on linux.

I had the same doubts, but the PNPDeviceID was the only working way I could get the Vendor/Product device data about the GPU.

If you prefer I can split GPU and PCI leaving the PCI to posterity

File: baseboard_windows.go (hunk 1 above)
Comment (jaypipes) [reply baseboard_windows.go]: Yeah, I think I'd prefer to just use a static const string query versus using an exported/public struct named the same as the WQL database table. It's just cleaner IMHO :)

File: pci_windows.go (hunk 27 above)
Comment (jaypipes) [reply pci_windows.go]: Yeah, let's split the GPU and PCI stuff for a separate PR where we can discuss this topic in isolation.

**Comment by KingRial** [comment]
Closing, since current request is going to be splitted in different pull requests
